lib/fencing/st_client.c

```diff
@@ -854,20 +854,23 @@ stonith_api_del_callback(stonith_t * stonith, int call_id, bool all_callbacks)
  * \param[in] st        Fencer API connection
  * \param[in] call_id   If positive, call ID of completed fence action, otherwise
  *                      legacy return code for early action failure
- * \param[in] rc        Legacy return code for action result
+ * \param[in] result    Full result for action
  * \param[in] userdata  User data to pass to callback
  * \param[in] callback  Fence action callback to invoke
  */
 static void
-invoke_fence_action_callback(stonith_t *st, int call_id, int rc, void *userdata,
+invoke_fence_action_callback(stonith_t *st, int call_id,
+                             pcmk__action_result_t *result,
+                             void *userdata,
                              void (*callback) (stonith_t *st,
                                                stonith_callback_data_t *data))
 {
     stonith_callback_data_t data = { 0, };
 
     data.call_id = call_id;
-    data.rc = rc;
+    data.rc = pcmk_rc2legacy(stonith__result2rc(result));
     data.userdata = userdata;
+    data.opaque = (void *) result;
```

Comment:
when we later add more opaque data we make it a structure that has result as the first member to keep it backwards compatible so that we don't have to deal with the struct for now - right?

Reply:
Exactly, except it won't even matter whether it's the first member because the contents will be internal, and the pointer can only be allocated by the fencing library of the same code base as the contents. The public API will only have access to the bare pointer, so the public struct won't change size. The pointer will only be accessible via methods so the internal implementation of the opaque data is flexible.

Reply:
I'm aware that code using the public API won't be able to access data directly.
Was more thinking about weird stuff the linker might be doing as to catch different library versions.
But I agree that there is probably no danger in this case unless sbdy really provokes it using dlopen(), dlsym() ...